content/learning-paths/smartphones-and-mobile/android_webgpu_dawn/6-render-a-simple-3D-object-part-2.md

```diff
@@ -0,0 +1,124 @@
+
+{{% notice Note %}}
+Exactly one of your source files must define `TINYOBJLOADER_IMPLEMENTATION` before including it:
+
+```C++
+#define TINYOBJLOADER_IMPLEMENTATION // add this to exactly 1 of your C++ files
+#include "tiny_obj_loader.h"
+```
+
+{{% /notice %}}
+
+We have developed a helper function [`loadGeometryFromObj`](https://github.com/varunchariArm/Android_DawnWebGPU/blob/main/app/src/main/cpp/webgpuRenderer.cpp#L475) using the above library.
+
+With all the setup done, we are now ready to render our 3D object
+
+## Rendering using WebGPU
+
+Now that we have done all the setup, we can run a rendering pass and *draw* something onto our *surface*. To encode any commands to be issued to GPU, we need to create a `CommandEncoder`, which is done:
+
+```C++
+wgpu::CommandEncoderDescriptor commandEncoderDesc;
+commandEncoderDesc.label = "Command Encoder";
+wgpu::CommandEncoder encoder = device.createCommandEncoder(commandEncoderDesc);
+```
+
+Next step is to create `RenderPassEncoder` using `encoder.beginRenderPass()` API:
+
+```C++
+wgpu::RenderPassDescriptor renderPassDesc{};
+
+wgpu::RenderPassColorAttachment renderPassColorAttachment{};
+renderPassColorAttachment.view = nextTexture;
+renderPassColorAttachment.resolveTarget = nullptr;
+renderPassColorAttachment.loadOp = wgpu::LoadOp::Clear;
+renderPassColorAttachment.storeOp = wgpu::StoreOp::Store;
+renderPassColorAttachment.clearValue = Color{ 1, 1, 1, 1.0 };
+renderPassColorAttachment.depthSlice = WGPU_DEPTH_SLICE_UNDEFINED;
+renderPassDesc.colorAttachmentCount = 1;
+renderPassDesc.colorAttachments = &renderPassColorAttachment;
+renderPassDesc.timestampWrites = nullptr;
+wgpu::RenderPassEncoder renderPass = encoder.beginRenderPass(renderPassDesc);
+```
+
+{{% notice Note %}}
+`ColorAttachment` is the only mandatory field. Also make sure you have specified `renderPassColorAttachment.depthSlice`
+{{% /notice %}}
+
+Now we can invoke the following APIs to draw our 3D object:
+
+* `renderPass.setPipeline(...);`
+* `renderPass.setVertexBuffer(...)`
+* `renderPass.setBindGroup(...)`
+* `renderPass.draw(...)`
+
+To finish encoding the sequence of commands and issue them to the GPU, few more API calls are needed:
+
+* End render pass `renderPass.end()`
+* Finish the command
+  
+  ```C++
+  wgpu::CommandBufferDescriptor cmdBufferDescriptor{};
+  cmdBufferDescriptor.label = "Command buffer";
+  wgpu::CommandBuffer command = encoder.finish(cmdBufferDescriptor);
+  encoder.release();
+  ```
+
+* Submit the Queue `queue.submit(command)`
+* Present the object onto surface `surface_.present()`
+
+{{% notice Tip %}}
+Make sure you release the created encoders and buffers by calling the respective `.release()` in order to avoid dangling pointer or other errors.
+{{% /notice %}}
+
+Finally we have to make a call `device_.tick()`, so that the `callbacks` are called and the ensure proper behavior.
```

Comment:
`device.tick()` should be called implicitly from `queue.submit()`. It should only be necessary if you're eg. waiting for a buffer to become mapped or for debugging purposes.

Reply:
removed